### PR TITLE
skafos 1.0 (new formula)

### PR DESCRIPTION
**Pull request by wess**
- [ X ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [ X ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ X ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ X ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by wess** [comment]
When I remove lib archive from the Formula, CI can't build it saying "Can't find lib archive" when I add it back it fails with: `Dependency 'libarchive' may be unnecessary as it is provided by macOS; try to build this formula without it.`

I see other formula in Formulas have `depends_on "lib archive"` Im not sure what to do to pass CI so I can get this merged /cc @defunkt @pjhyett @kevinclark @technoweenie @kevwil

**Comment by ilovezfs** [comment]
You can use `libarchive` if it needs those headers.

`libgit2` looks unused.

File: Formula/skafos.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  bottle :unneeded
+
+  depends_on "cmake" => :build
+  depends_on "libgit2"
+  depends_on "yaml-cpp"
+  depends_on "libarchive"
+
+  def install
+    ENV.deparallelize
```
Comment (ilovezfs) [inline Formula/skafos.rb]: this should be removed

File: Formula/skafos.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+class Skafos < Formula
+  desc      "Command-line tool for working with the Metis Machine platform"
+  homepage  "https://metismachine.com"
+  url       "https://github.com/MetisMachine/skafos/archive/1.0.tar.gz"
+  sha256    "254500f6a5bd4bee3cbfc8a7d1c8ceec1595430d5b14a69d859f9de80b8ebd16"
+  head      "https://github.com/metismachine/skafos.git"
+
+  bottle :unneeded
```
Comment (ilovezfs) [inline Formula/skafos.rb]: if it runs `make`, it's not bottle unneeded.

File: Formula/skafos.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  def install
+    ENV.deparallelize
+
+    system "make", "homebrew"
+    prefix.install Dir["_build/*"]
+  end
+
+  test do
+    `system "#{bin}/skafos", "--version"`
```
Comment (ilovezfs) [inline Formula/skafos.rb]: The backticks aren't needed.

**Review by ilovezfs — CHANGES_REQUESTED** [review]
The test needs to exercise and verify the functionality of the software, not run --version or --help.

File: Formula/skafos.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+class Skafos < Formula
+  desc      "Command-line tool for working with the Metis Machine platform"
```
Comment (ilovezfs) [inline Formula/skafos.rb]: only 1 space

File: Formula/skafos.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  homepage  "https://metismachine.com"
```
Comment (ilovezfs) [inline Formula/skafos.rb]: trailing slash

File: Formula/skafos.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  url       "https://github.com/MetisMachine/skafos/archive/1.0.tar.gz"
+  sha256    "254500f6a5bd4bee3cbfc8a7d1c8ceec1595430d5b14a69d859f9de80b8ebd16"
+  head      "https://github.com/metismachine/skafos.git"
```
Comment (ilovezfs) [inline Formula/skafos.rb]: please remove

File: Formula/skafos.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+  bottle :unneeded
+
+  depends_on "cmake" => :build
+  depends_on "libgit2"
```
Comment (ilovezfs) [inline Formula/skafos.rb]: not used

File: Formula/skafos.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  depends_on "yaml-cpp"
```
Comment (ilovezfs) [inline Formula/skafos.rb]: sort the non-build deps alphabetically amongst themselves

**Comment by ilovezfs** [comment]
```
iMac-TMP:~ joe$ ls -1 /usr/local/Cellar/skafos/1.0/
CMakeCache.txt
CMakeFiles
CPackConfig.cmake
CPackSourceConfig.cmake
INSTALL_RECEIPT.json
LICENSE
Makefile
README.md
_CPack_Packages
bin
cmake_install.cmake
install_manifest.txt
lib
libs
skafos-1.0-Darwin.sh
skafos-1.0-Darwin.tar.gz
src
```
This layout is highly unusual and won't be properly linked into the HOMEBREW_PREFIX (/usr/local).

The only top level items that make sense there are
```
INSTALL_RECEIPT.json
LICENSE
README.md
bin
lib
```
(The first three are installed automatically)

The rest are fish out of water.

**Comment by ilovezfs** [comment]
For `cmake`, it should be sufficient to run
```
system "cmake", ".", *std_cmake_args
system "make", "install"
```
or
```
mkdir "build" do
  system "cmake", "..", *std_cmake_args
  system "make", "install"
end
```
when an in-tree build isn't allowed.

If neither of those are working, there's something off with the build system.

File: Formula/skafos.rb (hunk 1 above)
Comment (wess) [reply Formula/skafos.rb]: this is amazing, thank you so much

File: Formula/skafos.rb (hunk 6 above)
Comment (wess) [reply Formula/skafos.rb]: remove all of these?

Comment (ilovezfs) [reply Formula/skafos.rb]: just head

**Comment by wess** [comment]
@ilovezfs -- Still seeing this error from your CI with the changes you requested: http://wess.co/otNn

**Comment by ilovezfs** [comment]
@wess it's a --new-formula error. Once merged it won't be triggered even by --strict.

File: Formula/skafos.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  homepage "https://metismachine.com/"
+  url "https://github.com/MetisMachine/skafos/archive/1.0.tar.gz"
+  sha256 "254500f6a5bd4bee3cbfc8a7d1c8ceec1595430d5b14a69d859f9de80b8ebd16"
+
+  depends_on "cmake" => :build
+  depends_on "libarchive"
+  depends_on "yaml-cpp"
+
+  def install
+    system "make", "clean"
```
Comment (ilovezfs) [inline Formula/skafos.rb]: are you sure this is needed?

File: Formula/skafos.rb [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    system "make", "_create_version_h"
+    system "make", "_env_for_prod"
+
+    mkdir "_build" do
+      system "cmake", "..", *std_cmake_args
```
Comment (ilovezfs) [inline Formula/skafos.rb]: looks like an in-tree build works fine. so
```
system "cmake", ".", *std_cmake_args
system "make", "install"
```
without the mkdir "_build"

File: Formula/skafos.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+      system "make", "install"
+    end
+  end
+
+  test do
+    system "#{bin}/skafos", "--version"
```
Comment (ilovezfs) [inline Formula/skafos.rb]: still need a functional test down here, not just --version or --help

**Comment by ilovezfs** [comment]
what gets installed looks much better now.
```
iMac-TMP:1.0 joe$ tree
.
├── INSTALL_RECEIPT.json
├── LICENSE
├── README.md
├── bin
│   └── skafos
├── include
│   ├── docopt
│   │   ├── docopt.h
│   │   ├── docopt_private.h
│   │   ├── docopt_util.h
│   │   └── docopt_value.h
│   └── json11.hpp
└── lib
    ├── cmake
    │   └── docopt
    │       ├── docopt-config-version.cmake
    │       ├── docopt-config.cmake
    │       ├── docopt-targets-release.cmake
    │       └── docopt-targets.cmake
    ├── libdocopt.a
    ├── libdocopt.dylib
    ├── libjson11.a
    └── pkgconfig
        └── json11.pc

7 directories, 17 files
```

File: Formula/skafos.rb (hunk 11 above)
Comment (wess) [reply Formula/skafos.rb]: we don't have any unit tests (at the moment), and we need auth for other executions, so suggestions on what we could use for a test? It's a CLI like heroku's where the user needs to login to use features.

**Comment by ilovezfs** [comment]
```
iMac-TMP:1.0 joe$ otool -L lib/libdocopt.dylib 
lib/libdocopt.dylib:
	/usr/local/opt/skafos/lib/libdocopt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```
seems a bit odd that the dylib has no version information being set. Compare for example
```
bash-3.2$ ls -l /usr/lib/libz.*
lrwxr-xr-x  1 root  wheel      12 Dec 25  2015 /usr/lib/libz.1.1.3.dylib -> libz.1.dylib
lrwxr-xr-x  1 root  wheel      12 Dec 25  2015 /usr/lib/libz.1.2.5.dylib -> libz.1.dylib
-rwxr-xr-x  1 root  wheel  177728 Mar 12  2016 /usr/lib/libz.1.dylib
lrwxr-xr-x  1 root  wheel      12 Dec 25  2015 /usr/lib/libz.dylib -> libz.1.dylib
bash-3.2$ otool -L /usr/lib/libz.dylib 
/usr/lib/libz.dylib:
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
bash-3.2$ 
```

**Comment by wess** [comment]
@ilovezfs -- Thank you so much for all your help, I will be sure our company makes a donation, as it's well deserved.

**Comment by ilovezfs** [comment]
You're welcome @wess :)

**Comment by wess** [comment]
@ilovezfs libdocopt doesn't have a brew formula, and I just pulled master and cleaned up their cmake to use it in this

**Comment by ilovezfs** [comment]
@wess system zlib was just an example. You can see the same sort of version info for say libxml2

```
bash-3.2$ ls -l /usr/lib/libxml2.*
-rwxr-xr-x  1 root  wheel  2395360 Nov 14 22:18 /usr/lib/libxml2.2.dylib
lrwxr-xr-x  1 root  wheel       15 Dec 25  2015 /usr/lib/libxml2.dylib -> libxml2.2.dylib
bash-3.2$ otool -L /usr/lib/libxml2.dylib 
/usr/lib/libxml2.dylib:
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/usr/lib/libicucore.A.dylib (compatibility version 1.0.0, current version 55.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
bash-3.2$
```
See the `	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)`

as opposed to

```
/usr/local/opt/skafos/lib/libdocopt.dylib (compatibility version 0.0.0, current version 0.0.0)
```

**Comment by wess** [comment]
@ilovezfs right, which is why I was telling you I pulled docopt from their master in order for it to work, and they must not version their master.

**Comment by ilovezfs** [comment]
Ah I see what you mean.

**Comment by wess** [comment]
Im just trying to figure out what I can run as a test, for the time being. help and version run through the pipeline of the app, testing auth and such, which is why I used them

**Comment by ilovezfs** [comment]
```
    (testpath/"test.exp").write <<~EOS
      spawn #{bin}/skafos setup
      set timeout 5
      expect {
        timeout { exit 1 }
        "Please enter email:"
      }
      send "me@foo.bar\r"
      expect {
        timeout { exit 2 }
        "Please enter password:"
      }
      send "1234\r"
      expect {
        timeout { exit 3 }
        eof
      }
    EOS
    assert_match "Invalid email or password", shell_output("expect -f test.exp")
  end
```
maybe?

**Comment by wess** [comment]
hah, I didn't think of writing the test in ruby right there.

**Comment by wess** [comment]
So I added your suggested test in there, but I don't see in the docs how I can run just the test to see if it works when I run `brew install --build-from-source skafos --debug --verbose`

**Comment by ilovezfs** [comment]
Also, I'm wondering if this should be linking `libdocopt.a` and not installing anything except `bin/skafos`

**Comment by ilovezfs** [comment]
To run the test, `brew test -vd foo`

**Comment by ilovezfs** [comment]
As it is, this is dependent on the prefix being `/usr/local` and won't work for users who use an alternate prefix.
```
iMac-TMP:~ joe$ /usr/local/Cellar/skafos/1.0/bin/skafos -v

Skafos version: 1.0-dev
Metis Machine https://metismachine.com

iMac-TMP:~ joe$ brew unlink skafos
Unlinking /usr/local/Cellar/skafos/1.0... 8 symlinks removed
iMac-TMP:~ joe$ /usr/local/Cellar/skafos/1.0/bin/skafos -v
dyld: Library not loaded: @rpath/libdocopt.dylib
  Referenced from: /usr/local/Cellar/skafos/1.0/bin/skafos
  Reason: image not found
Trace/BPT trap: 5
iMac-TMP:~ joe$ 
```

**Comment by wess** [comment]
are you in the U.S.?

**Comment by ilovezfs** [comment]
yes

**Comment by wess** [comment]
does home-brew have an irc channel or slack, that way you don't have to have your browser ping you constantly :)

**Comment by ilovezfs** [comment]
`#machomebrew` on freenode is easiest

**Comment by ilovezfs** [comment]
```
iMac-TMP:1.0 joe$ tree .
.
├── INSTALL_RECEIPT.json
├── LICENSE
├── README.md
├── bin
│   └── skafos
├── include
│   └── json11.hpp
└── lib
    ├── libjson11.a
    └── pkgconfig
        └── json11.pc

4 directories, 7 files
```
probably should remove the libjson stuff from the `make install` target too.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Thanks for your first contribution to Homebrew, @wess! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**